segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/hdfs/HDFSStorage.java

```diff
@@ -408,7 +412,8 @@ public SegmentHandle openWrite(String streamSegmentName) throws StreamSegmentExc
             } catch (IOException e) {
                 throw HDFSExceptionHelpers.convertException(streamSegmentName, e);
             }
-        } while (fencedCount < MAX_ATTEMPT_COUNT);
+            // Looping for the maximum possible number.
+        } while (fencedCount < MAX_EPOCH);
```

Comment:
This opens up the segment store to a loop that can severely affect availability. If everything works as expected, then we will surely do very few iterations, but it is not good practice to enable a loop that can go on for a very large number of iterations.

If you guys don't want to use the cluster size, possibly because it changes dynamically, then we need to think of another bound. Another option is that we add a path for the segment store to tell the controller that it is giving up on a segment store so that the controller performs another assignment, possibly to the same segment store instance, and the segment container does not remain orphaned. See it as a graceful shutdown of the segment container.

Reply:
I concur with the worry about the size of the loop.
Here are some caveats:
1. This is segment specific: Availability of a single segment is affected and not the whole segmentstore.
2. We do not have any scenario where this loop goes more than total number of segmentstore instances for the winner.
So if we design any communication from storage layer to segment store on the failure path, it should be segment specific. 
I think another option can be to limit this loop to a high enough number (greater than maximum possible segmentstore instances  - say 500/5000).

Reply:
The verification is being performed on a per segment basis, but this could happen with a batch of segments concurrently because ownership is transferred at the granularity of segment containers.

Is there anything we can do with epochs? Not the `MAX_EPOCH` value, but with the real epoch values observed.

Reply:
ok. This is a good idea. Each segmentstore can loop at the max for its own epoch and not necessarily for the max epoch. If it is the current owner, the ownership should be resolved by this time. Will update the PR for this.